src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/RSAPssParamAsn.cs

```diff
@@ -0,0 +1,51 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.Asn1;
+
+namespace System.Security.Cryptography.X509Certificates.Asn1
+{
+    // https://tools.ietf.org/html/rfc4055#section-3.1
+    //
+    // RSASSA-PSS-params ::= SEQUENCE {
+    //     hashAlgorithm      [0] HashAlgorithm DEFAULT sha1Identifier,
+    //     maskGenAlgorithm   [1] MaskGenAlgorithm DEFAULT mgf1SHA1Identifier,
+    //     saltLength         [2] INTEGER DEFAULT 20,
+    //     trailerField       [3] INTEGER DEFAULT 1
+    // }
+    //
+    // mgf1SHA1Identifier  AlgorithmIdentifier  ::= { id-mgf1, sha1Identifier }
+    // sha1Identifier  AlgorithmIdentifier  ::=  { id-sha1, NULL }
+    // (and similar for SHA256/384/512)
+    //
+    // RFC 5754 says that the NULL for SHA2 (256/384/512) MUST be omitted
+    // (https://tools.ietf.org/html/rfc5754#section-2) (and that you MUST
+    // be able to read it even if someone wrote it down)
+    //
+    // Since we
+    //  * don't support SHA-1 in this class
+    //  * only support MGF-1
+    //  * don't support the MGF PRF being different than hashAlgorithm
+    //  * use saltLength==hashLength
+    //  * don't allow custom trailer
+    // we don't have to worry about any of the DEFAULTs. (specify, specify, specify, omit).
```

Comment:
This comment doesn't make sense now that this is visible outside that one consumer.

Since this could now get easily picked up by some other component, please specify the DefaultValue values for `hashAlgorithm` and `maskGenAlgorithm`

`hashAlgorithm`'s default is (I think) `30 09 06 05 2B 0E 03 02 1A 05 00` and `maskGenAlgorithm`'s is `30 16 06 09 2A 86 48 86 F7 0D 01 01 08 30 09 06 05 2B 0E 03 02 1A 05 00`.

At least, based on the comments on lines 19 and 20.

The comment for lines 23-25 seem to still belong at the caller, about how to build the AlgorithmIdentifierAsn.

Reply:
I think the default values are just 30 07 06 05 2B 0E 03 02 1A and 30 14 06 09 2A 86 48 86 F7 0D 01 01 08 30 07 06 05 2B 0E 03 02 1A (without the NULLs at the end).

Reply:
Omitting the NULL is correct for the SHA-2 algorithms, but https://tools.ietf.org/html/rfc3447 says

```ASN.1
sha1    HashAlgorithm ::= {
    algorithm   id-sha1,
    parameters  SHA1Parameters : NULL  -- included for compatibility
                                       -- with existing implementations
}

mgf1SHA1    MaskGenAlgorithm ::= {
    algorithm   id-mgf1,
    parameters  HashAlgorithm : sha1
}

--
-- AlgorithmIdentifier.parameters for id-RSASSA-PSS.
-- Note that the tags in this Sequence are explicit.
--
RSASSA-PSS-params ::= SEQUENCE {
    hashAlgorithm      [0] HashAlgorithm      DEFAULT sha1,
    maskGenAlgorithm   [1] MaskGenAlgorithm   DEFAULT mgf1SHA1,
    saltLength         [2] INTEGER            DEFAULT 20,
    trailerField       [3] TrailerField       DEFAULT trailerFieldBC
}
```

So the NULLs are semantically within the SHA-1-based default values.

Reply:
I based it on the default value for Version in TbsCertificate.  Looking at the serializer code I think it does.

I don't see any specific tests for serializing and deserializing default explicit values in the encoding tests library, seems like it would be good to add one so we can say for certain :smile:.

Reply:
The default value of Version in TbsCertificate is currently not used in practice or any of the tests afaik (and it was copied from the older Pkcs code). The default values of RSAPssParamAsn for the last two fields are tested to some extent though (`RSAPssX509SignatureGeneratorTests.SignatureAlgorithm_Encoding`). I will make a test for it.

Reply:
> I will make a test for it.

Yay :smile:.

I second-, third-, fourth-, and fifth- guessed myself as to whether the default values reading and writing was inside or outside the explicit tag.

Reply:
Interesting, there's one more copy of `TbsCertificate` in `X509DeserializeTests` and that one has the `DefaultValue` without the explicit tag. :-) It still doesn't use the default value in the tests though.

Reply:
That would explain why I kept n-th guessing myself.  It probably just hasn't come up in a read/write scenario yet.

The smoothest approach is probably to fix that aspect of the serializer (with tests) in a separate PR, and then resume this.  Otherwise, make sure that nothing is doing both reading and writing of the same structure which has default+explicit fields.  (AppleCertificatePal is reading TbsCertificate.Version, CertificateRequest is writing it, so they can't both go in together)

Reply:
I will do more tests tomorrow to see if I am able to fix it.

Reply:
Ok, I submitted PR for the deserializer along with unit tests. Now I noticed that there's already PssParamsAsn in the Pkcs code. Will reuse that one after I rebase this PR :-)

Reply:
...ouch, but that one also has the default values without the explicit NULL at the end.